provider/pkg/schema/extension_resource.go

```diff
@@ -0,0 +1,272 @@
+	}
+}
+
+func (e *ExtensionResource) ToOutputs(outputs map[string]any) resource.PropertyMap {
+	return resource.NewPropertyMapFromMap(map[string]any{
+		"type":         e.Type,
+		"properties":   e.Properties,
+		"writeOnly":    e.WriteOnly,
+		"tagsProperty": e.TagsProperty,
+		"tagsStyle":    e.TagsStyle,
+		"outputs":      outputs,
+	})
+}
+
+type extensionResource struct {
+	host   *provider.HostClient
+	client client.CloudControlClient
+}
+
+func NewExtensionResource(host *provider.HostClient, client client.CloudControlClient) *extensionResource {
+	return &extensionResource{
+		host:   host,
+		client: client,
+	}
+}
+
+func (r *extensionResource) Create(ctx context.Context, urn resource.URN, inputs resource.PropertyMap) (identifier *string, outputs resource.PropertyMap, err error) {
+	var extensionResource ExtensionResource
+	_, err = resourcex.Unmarshal(&extensionResource, inputs, resourcex.UnmarshalOptions{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal inputs: %w", err)
+	}
+
+	id, resourceState, err := r.client.Create(ctx, extensionResource.Type, extensionResource.Properties)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create resource: %w", err)
+	}
+
+	return id, extensionResource.ToOutputs(resourceState), nil
+}
+
+func (r *extensionResource) Read(ctx context.Context, urn resource.URN, id string, oldInputs, oldState resource.PropertyMap) (resource.PropertyMap, resource.PropertyMap, bool, error) {
+	if len(oldState) == 0 {
+		// We can't yet support import because the type would not be known.
+		return nil, nil, false, fmt.Errorf("ExtensionResource import not implemented")
+	}
+	var extensionResource ExtensionResource
+	_, err := resourcex.Unmarshal(&extensionResource, oldInputs, resourcex.UnmarshalOptions{})
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("failed to unmarshal old inputs: %w", err)
+	}
+
+	var oldOutputsMap resource.PropertyMap
+	if oldOutputsValue, ok := oldState[resource.PropertyKey("outputs")]; ok && oldOutputsValue.IsObject() {
+		oldOutputsMap = oldOutputsValue.ObjectValue()
+	}
+
+	resourceState, exists, err := r.client.Read(ctx, extensionResource.Type, id)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("failed to read resource: %w", err)
+	}
+	if !exists {
+		return nil, nil, false, nil
+	}
+
+	// Fill in the write-only properties from the old state as they won't included when reading.
+	if len(extensionResource.WriteOnly) > 0 {
+		missingProps := make([]string, 0, len(extensionResource.WriteOnly))
+		for _, writeOnlyProp := range extensionResource.WriteOnly {
+			if _, ok := resourceState[writeOnlyProp]; !ok {
+				oldValue, ok := extensionResource.Properties[writeOnlyProp]
+				missingProps = append(missingProps, writeOnlyProp)
+				if ok {
+					resourceState[writeOnlyProp] = oldValue
+				}
+			}
+		}
+		r.host.Log(ctx, diag.Warning, urn, fmt.Sprintf("Can't refresh write-only properties: %s", strings.Join(missingProps, ", ")))
```

Comment:
Why is it even a warning if we can't refresh a write-only property? Is it not true by definition?

Reply:
This is also already the behaviour of all other resources. We inform the user during a refresh, which properties won't have been refreshed in case they were expecting to see a change but didn't. Perhaps we could downgrade it to info.

Reply:
I guess, perhaps because the user has specified this metadata themselves, we could skip this message here. Was that what you were suggesting?